core/optaplanner-core-impl/src/test/java/org/optaplanner/core/impl/solver/change/DefaultProblemChangeDirectorTest.java

```diff
@@ -96,4 +103,68 @@ void complexProblemChange_correctlyNotifiesScoreDirector() {
         verify(scoreDirectorMock, times(1)).beforeProblemFactRemoved(removedFact);
         verify(scoreDirectorMock, times(1)).afterProblemFactRemoved(removedFact);
     }
+
+    @Test
+    void initScoreUpdateAfterEntityAddedRemoved() {
+        InnerScoreDirector<TestdataListSolution, SimpleScore> scoreDirector =
+                PlannerTestUtils.mockScoreDirectorWithLookupEnabled(TestdataListSolution.buildSolutionDescriptor());
+
+        TestdataListSolution solution = TestdataListSolution.generateInitializedSolution(6, 2);
+        scoreDirector.setWorkingSolution(solution);
+        DefaultProblemChangeDirector<TestdataListSolution> problemChangeDirector =
+                new DefaultProblemChangeDirector<>(scoreDirector);
+
+        TestdataListValue value = new TestdataListValue("v");
+
+        ProblemChange<TestdataListSolution> addValue =
+                (workingSolution, director) -> director.addEntity(value, workingSolution::addValue);
+
+        assertThat(problemChangeDirector.doProblemChange(addValue).getInitScore()).isEqualTo(-1);
+
+        ProblemChange<TestdataListSolution> removeValue =
+                (workingSolution, director) -> director.removeEntity(value, workingSolution::removeValue);
+
+        assertThat(problemChangeDirector.doProblemChange(removeValue).getInitScore()).isEqualTo(0);
+
+        // Removing an assigned value doesn't impact the init score.
+        TestdataListValue assignedValue = solution.getEntityList().get(0).getValueList().get(0);
+        ProblemChange<TestdataListSolution> removeAssignedValue =
+                (workingSolution, director) -> director.removeEntity(assignedValue, v -> {
+                    workingSolution.getEntityList().get(0).getValueList().remove(v);
+                    workingSolution.removeValue(v);
+                });
+        assertThat(problemChangeDirector.doProblemChange(removeAssignedValue).getInitScore()).isEqualTo(0);
+    }
+
+    @Test
+    void initScoreUpdateAfterFactAddedRemoved() {
+        InnerScoreDirector<TestdataListSolutionExternalized, SimpleScore> scoreDirector =
+                PlannerTestUtils.mockScoreDirectorWithLookupEnabled(TestdataListSolutionExternalized.buildSolutionDescriptor());
+
+        TestdataListSolutionExternalized solution = TestdataListSolutionExternalized.generateInitializedSolution(3, 1);
+        scoreDirector.setWorkingSolution(solution);
+        DefaultProblemChangeDirector<TestdataListSolutionExternalized> problemChangeDirector =
+                new DefaultProblemChangeDirector<>(scoreDirector);
+
+        TestdataListValueExternalized value = new TestdataListValueExternalized("v");
+
+        ProblemChange<TestdataListSolutionExternalized> addValue =
+                (workingSolution, director) -> director.addProblemFact(value, workingSolution::addValue);
+
+        assertThat(problemChangeDirector.doProblemChange(addValue).getInitScore()).isEqualTo(-1);
```

Comment:
This is why the test is invalid. After the last PC in the queue is done, solver is restarted. That involves `AbstractScoreDirector.setWorkingSolution()` where the init score is calculated from scratch. Therefore, it is redundant to keep the init score up-to-date after each individual PC.

This test was written in anticipation of truly incremental PCs. But that turned out to be impossible because solver needs to override the working solution with `setWorkingSolutionFromBestSolution()` before the first PC in a batch and consequently _restart_ and run through all phases after the last PC.